### PR TITLE
Support Azure OpenAI and AI Search

### PR DESCRIPTION
**Pull request by showpune**
Add starter for 
Azure Open AI, include
1.  AzureOpenAiChatModel
2.  AzureOpenAiStreamingChatModel
3.  AzureOpenAiEmbeddingModel
4.  AzureOpenAiImageModel
Add support for AI Search ( dependen on PR https://github.com/langchain4j/langchain4j/pull/758)

Need the env

> AZURE_OPENAI_KEY
> AZURE_OPENAI_ENDPOINT
> AZURE_OPENAI_DEPLOYMENT_NAME
> AZURE_EMBEDDING_DEPLOYMENT_NAME
> AZURE_AISEARCH_API_KEY
> AZURE_AISEARCH_ENDPOINT
> AZURE_AISEARCH_SETUP_INDEX
> AZURE_AISEARCH_DIMENSIONS

To run the IT

### CODE REVIEW

**Review by langchain4j — COMMENTED** [review]
@showpune thanks a lot!

File: langchain4j-azure-open-ai-spring-boot-starter/pom.xml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+        <!-- should be listed before spring-boot-configuration-processor -->
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-core</artifactId>
```
[the remark this answers is not in the document]
Comment (showpune) [reply langchain4j-azure-open-ai-spring-boot-starter/pom.xml]: Removed

File: langchain4j-azure-spring-boot-starter/pom.xml [hunk 2]
```diff
@@ -22,6 +22,11 @@
             <artifactId>langchain4j-azure-open-ai</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-azure-ai-search</artifactId>
```
Comment (langchain4j) [inline langchain4j-azure-spring-boot-starter/pom.xml:27]: Users of langchain4j-azure-open-ai will not necessarily use langchain4j-azure-ai-search, so I would avoid putting them together. Also, langchain4j-azure-ai-search will have some changes soon based on [this PR](https://github.com/langchain4j/langchain4j/pull/587), so I would wait with creating a starter before things are settled in its implementation.

Comment (showpune) [reply langchain4j-azure-spring-boot-starter/pom.xml:27]: I changed the starter name to langchain4j-azure-spring-boot-starter as I want to put the related resource like openai, ai search into one starter, as if the customer is an azure customer, multiple service will be used togethter.

Comment (jdubois) [reply langchain4j-azure-spring-boot-starter/pom.xml:27]: I would split them in `langchain4j-azure-openai-spring-boot-starter` and `langchain4j-azure-ai-search-spring-boot-starter`, so people can use on or the other without having the dependencies coming from the one they don't use.
@showpune I believe we work on the same thing here, let me ping you internally 😀

Comment (langchain4j) [reply langchain4j-azure-spring-boot-starter/pom.xml:27]: Agree, there are 2 separate modules in langchain4j, so it makes more sense to have 2 separate starters

Comment (showpune) [reply langchain4j-azure-spring-boot-starter/pom.xml:27]: Sure.

Wait for https://github.com/langchain4j/langchain4j/pull/796 merge

Comment (langchain4j) [reply langchain4j-azure-spring-boot-starter/pom.xml:27]: @showpune merged!

Comment (showpune) [reply langchain4j-azure-spring-boot-starter/pom.xml]: Ready for release

File: langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/Properties.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+public class Properties {
+
+    static final String PREFIX = "langchain4j.azure.ai-search";
+
+    String endpoint;
+    String apiKey;
+    int dimensions;
+    boolean createOrUpdateIndex;
+    int maxResults = 1;
+    double minScore = 0.6;
```
Comment (jdubois) [inline langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/Properties.java:20]: The `DEFAULT_MIN_SCORE` constant is at `0.0` so I would keep it to avoid confusion

File: langchain4j-azure-aisearch-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    private static final String AZURE_AISEARCH_SETUP_INDEX = System.getenv("AZURE_AISEARCH_SETUP_INDEX");
+    private static final String AZURE_AISEARCH_MAX_RESULTS = System.getenv("AZURE_AISEARCH_MAX_RESULTS");
+    private static final String AZURE_AISEARCH_MIN_SCORE = System.getenv("AZURE_AISEARCH_MIN_SCORE");
+    private static final String AZURE_AISEARCH_QUERY_TYPE = System.getenv("AZURE_AISEARCH_QUERY_TYPE");
+
+
+    ApplicationContextRunner contextRunner = new ApplicationContextRunner()
+            .withConfiguration(AutoConfigurations.of(AutoConfig.class));
+
+//    @Test
```
Comment (jdubois) [inline langchain4j-azure-aisearch-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java:27]: You should not have a commented out test

File: langchain4j-azure-openai-spring-boot-starter/src/main/java/dev/langchain4j/azure/openai/spring/AutoConfig.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import org.springframework.context.annotation.Bean;
+
+import static dev.langchain4j.azure.openai.spring.Properties.PREFIX;
+
+@AutoConfiguration
+@EnableConfigurationProperties(Properties.class)
+public class AutoConfig {
+
+    @Bean
+    @ConditionalOnProperty(PREFIX + ".chat-model.api-key")
```
Comment (jdubois) [inline langchain4j-azure-openai-spring-boot-starter/src/main/java/dev/langchain4j/azure/openai/spring/AutoConfig.java:21]: This will only work for using Azure OpenAI AND the API Key:
- we should also support using OpenAI (see the `nonAzureApiKey` method)
- I'm not sure about supporting Entra ID as that becomes too complicated for a YAML file at some point, so let's not do this yet

This also means that you should check here that either `api-key` or `non-azure-api-key` exists.

File: langchain4j-azure-openai-spring-boot-starter/src/main/java/dev/langchain4j/azure/openai/spring/AutoConfig.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                .frequencyPenalty(chatModelProperties.getFrequencyPenalty())
+                .timeout(chatModelProperties.getTimeout())
+                .maxRetries(chatModelProperties.getMaxRetries())
+                .proxyOptions(ProxyOptions.fromConfiguration(Configuration.getGlobalConfiguration()))
+                .logRequestsAndResponses(chatModelProperties.getLogRequestsAndResponses() != null && chatModelProperties.getLogRequestsAndResponses())
+                .build();
+    }
+
+    @Bean
+    @ConditionalOnProperty(PREFIX + ".streaming-chat-model.api-key")
```
Comment (jdubois) [inline langchain4j-azure-openai-spring-boot-starter/src/main/java/dev/langchain4j/azure/openai/spring/AutoConfig.java:42]: Same as above about the `nonAzureApiKey`

File: langchain4j-azure-openai-spring-boot-starter/src/main/java/dev/langchain4j/azure/openai/spring/AutoConfig.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                .presencePenalty(chatModelProperties.getPresencePenalty())
+                .frequencyPenalty(chatModelProperties.getFrequencyPenalty())
+                .timeout(chatModelProperties.getTimeout())
+                .proxyOptions(ProxyOptions.fromConfiguration(Configuration.getGlobalConfiguration()))
+                .logRequestsAndResponses(chatModelProperties.getLogRequestsAndResponses() != null && chatModelProperties.getLogRequestsAndResponses())
+                .build();
+    }
+
+    @Bean
+    @ConditionalOnProperty(PREFIX + ".embedding-model.api-key")
```
Comment (jdubois) [inline langchain4j-azure-openai-spring-boot-starter/src/main/java/dev/langchain4j/azure/openai/spring/AutoConfig.java:62]: Same as above about the `nonAzureApiKey`

File: langchain4j-azure-openai-spring-boot-starter/src/main/java/dev/langchain4j/azure/openai/spring/AutoConfig.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                .deploymentName(embeddingModelProperties.getDeploymentName())
+                .timeout(embeddingModelProperties.getTimeout())
+                .maxRetries(embeddingModelProperties.getMaxRetries())
+                .proxyOptions(ProxyOptions.fromConfiguration(Configuration.getGlobalConfiguration()))
+                .logRequestsAndResponses(embeddingModelProperties.getLogRequestsAndResponses() != null && embeddingModelProperties.getLogRequestsAndResponses())
+                .build();
+    }
+
+    @Bean
+    @ConditionalOnProperty(PREFIX + ".image-model.api-key")
```
Comment (jdubois) [inline langchain4j-azure-openai-spring-boot-starter/src/main/java/dev/langchain4j/azure/openai/spring/AutoConfig.java:77]: Same as above about the `nonAzureApiKey`

**Review by jdubois — COMMENTED** [review]
Checking for the "deployment name" will also only work for Azure OpenAI and not for "non-Azure" OpenAI.
Can you test doing like at https://github.com/spring-projects/spring-ai/blob/ed68dfb00cac650b6c2b5e17f179e624a0f840d2/spring-ai-spring-boot-autoconfigure/src/main/java/org/springframework/ai/autoconfigure/azure/openai/AzureOpenAiChatProperties.java#L22 and enable the configuration when any property is available? To be honest I've never done that before, but that seems like a good option.

**Comment by showpune** [comment]
Spring don't support or condition, I updated the code using other solution

**Review by jdubois — APPROVED** [review]
It's sad that you need to have both configurations, but indeed that's the only way to have this working.
And 99,99% of people will use "API Key", and this configuration is the most logical and easiest for them.

File: pom.xml [hunk 9]
```diff
@@ -19,6 +19,8 @@
         <module>langchain4j-anthropic-spring-boot-starter</module>
         <module>langchain4j-ollama-spring-boot-starter</module>
         <module>langchain4j-open-ai-spring-boot-starter</module>
+        <module>langchain4j-azure-aisearch-spring-boot-starter</module>
```
Comment (langchain4j) [inline pom.xml:22]:
```suggestion
        <module>langchain4j-azure-ai-search-spring-boot-starter</module>
```

File: pom.xml [hunk 10]
```diff
@@ -19,6 +19,8 @@
...
+        <module>langchain4j-azure-openai-spring-boot-starter</module>
```
Comment (langchain4j) [inline pom.xml:23]:
```suggestion
        <module>langchain4j-azure-open-ai-spring-boot-starter</module>
```

File: langchain4j-azure-aisearch-spring-boot-starter/pom.xml [hunk 11]
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-spring</artifactId>
+        <version>0.29.0-SNAPSHOT</version>
```
Comment (langchain4j) [inline langchain4j-azure-aisearch-spring-boot-starter/pom.xml:10]:
```suggestion
        <version>0.29.0</version>
```

File: langchain4j-azure-aisearch-spring-boot-starter/pom.xml [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>langchain4j-azure-aisearch-spring-boot-starter</artifactId>
```
Comment (langchain4j) [inline langchain4j-azure-aisearch-spring-boot-starter/pom.xml:14]:
```suggestion
    <artifactId>langchain4j-azure-ai-search-spring-boot-starter</artifactId>
```

File: langchain4j-azure-aisearch-spring-boot-starter/pom.xml [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-embeddings-all-minilm-l6-v2</artifactId>
+            <version>0.29.0-SNAPSHOT</version>
```
Comment (langchain4j) [inline langchain4j-azure-aisearch-spring-boot-starter/pom.xml:59]:
```suggestion
            <version>0.29.0</version>
```

File: langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/AutoConfig.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+import static dev.langchain4j.azure.aisearch.spring.Properties.PREFIX;
+
+@AutoConfiguration
+@EnableConfigurationProperties(Properties.class)
+public class AutoConfig {
+
+    @Bean
+    @ConditionalOnProperty(PREFIX + ".api-key")
+    public AzureAiSearchContentRetriever azureAiSearchContentRetriever(Properties properties, EmbeddingModel embeddingModel) {
```
Comment (langchain4j) [inline langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/AutoConfig.java:18]: Perharps `EmbeddingModel` should be optional? It won't be present for a full-text search, right?

File: langchain4j-azure-openai-spring-boot-starter/pom.xml [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-spring</artifactId>
+        <version>0.29.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>langchain4j-azure-openai-spring-boot-starter</artifactId>
```
Comment (langchain4j) [inline langchain4j-azure-openai-spring-boot-starter/pom.xml:14]:
```suggestion
    <artifactId>langchain4j-azure-open-ai-spring-boot-starter</artifactId>
```

File: langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/AutoConfig.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+
+import static dev.langchain4j.azure.aisearch.spring.Properties.PREFIX;
+
+@AutoConfiguration
+@EnableConfigurationProperties(Properties.class)
+public class AutoConfig {
+
```
Comment (langchain4j) [inline langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/AutoConfig.java:15]: What about `AzureAiSearchEmbeddingStore`? Should we provide auto-config for it as well?

File: langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/Properties.java [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+@ConfigurationProperties(prefix = Properties.PREFIX)
+public class Properties {
+
+    static final String PREFIX = "langchain4j.azure.ai-search";
+
+    String endpoint;
+    String apiKey;
+    int dimensions;
+    boolean createOrUpdateIndex;
+    int maxResults = 1;
```
Comment (langchain4j) [inline langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/Properties.java:19]:
```suggestion
    int maxResults = 3;
```
We use "3" as a default value everywhere. Is there a reason why we should use 1 here?

File: langchain4j-azure-openai-spring-boot-starter/pom.xml [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-open-ai</artifactId>
```
Comment (langchain4j) [inline langchain4j-azure-openai-spring-boot-starter/pom.xml:59]: Why is this required?

**Review by langchain4j — APPROVED** [review]
@showpune thanks a lot!

File: langchain4j-azure-openai-spring-boot-starter/pom.xml (hunk 18 above)
Comment (showpune) [reply langchain4j-azure-openai-spring-boot-starter/pom.xml]: It was used for test and now replaced with ExampleTestTokenizer

File: langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/AutoConfig.java (hunk 16 above)
Comment (showpune) [reply langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/AutoConfig.java]: Added it

File: langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/AutoConfig.java (hunk 14 above)
Comment (showpune) [reply langchain4j-azure-aisearch-spring-boot-starter/src/main/java/dev/langchain4j/azure/aisearch/spring/AutoConfig.java]: Solved

**Review by jdubois — APPROVED** [review]

**Comment by showpune** [comment]
@langchain4j , can you help to approve if no more comments?

File: langchain4j-azure-ai-search-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+                ).run(context -> {
+                    ContentRetriever contentRetriever = context.getBean(ContentRetriever.class);
+                    assertThat(contentRetriever).isInstanceOf(AzureAiSearchContentRetriever.class);
+
+                    assertThat(context.getBean(AzureAiSearchContentRetriever.class)).isSameAs(contentRetriever);
+                });
+    }
+
+    @Test
+    void should_provide_ai_search_retrive_create_or_update_indexr() {
```
[the remark this answers is not in the document]
Comment (showpune) [reply langchain4j-azure-ai-search-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java:43]: Yes and if the create-or-update-index is setup to true, it create a new index. Or it just use the existing one

File: langchain4j-azure-ai-search-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+import dev.langchain4j.store.embedding.azure.search.AzureAiSearchEmbeddingStore;
+import org.junit.jupiter.api.Test;
+import org.springframework.boot.autoconfigure.AutoConfigurations;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class AutoConfigIT {
+
+    private static final String AZURE_AISEARCH_API_KEY = System.getenv("AZURE_AISEARCH_API_KEY");
```
[the remark this answers is not in the document]
Comment (showpune) [reply langchain4j-azure-ai-search-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java:17]: The component is called AISEARCH, what the meaning of main module?  What is the AZURE_SEARCH_KEY is mapped in main module?

[the remark this answers is not in the document]
Comment (langchain4j) [reply langchain4j-azure-ai-search-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java:17]: Sorry for unclarity, main module means https://github.com/langchain4j/langchain4j
`AZURE_SEARCH_KEY` env variable is used there in multiple places: in ITs, in env variables for builds, in repository secrets, etc.
So it would be nice if you could use it instead of forcing contributors (e.g. @jdubois and me) to configure the new env variable.
Thank you!

File: langchain4j-azure-ai-search-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (langchain4j) [reply langchain4j-azure-ai-search-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java:43]: To repeat myself:

> if I change it to false or remove this property completely, the test still works

In my understanding it means that either there is a bug in the test or the test does not test what it is supposed to test or the test name should not mention this property which does not affect the test.

[the remark this answers is not in the document]
Comment (showpune) [reply langchain4j-azure-ai-search-spring-boot-starter/src/test/java/dev/langchain4j/azure/aisearch/spring/AutoConfigIT.java]: I update the search test case

1. Using create-or-update-index=true to create a index
2. Using create-or-update-index=false to search, with 4 search types

File: langchain4j-azure-open-ai-spring-boot-starter/src/test/java/dev/langchain4j/azure/openai/AutoConfigIT.java [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class AutoConfigIT {
+
+    private static final String AZURE_OPENAI_KEY = System.getenv("AZURE_OPENAI_KEY");
+    private static final String AZURE_OPENAI_ENDPOINT = System.getenv("AZURE_OPENAI_ENDPOINT");
+    private static final String AZURE_OPENAI_DEPLOYMENT_NAME = System.getenv("AZURE_OPENAI_DEPLOYMENT_NAME");
```
[the remark this answers is not in the document]
Comment (showpune) [reply langchain4j-azure-open-ai-spring-boot-starter/src/test/java/dev/langchain4j/azure/openai/AutoConfigIT.java]: I update the code and using the fixed name deployment like in main module, is it OK?

**Review by jdubois — APPROVED** [review]

**Comment by showpune** [comment]
@langchain4j , can you help to check?

**Comment by langchain4j** [comment]
@showpune sure, sorry for delays, I have too much on my plate right now
